File: src/exports/LandRegistry.cc
#include "pland/Global.h"
#include "pland/PLand.h"
#include "pland/Version.h"
#include "pland/aabb/LandAABB.h"
#include "pland/land/Land.h"
#include "pland/utils/JSON.h"
#include <algorithm>
#include <unordered_map>
#include <vector>

#include "exports/APIHelper.h"

#include "ExportDef.h"

namespace ldapi {


void Export_Class_LandRegistry() {
    exportAs("LandRegistry_isOperator", [](std::string const& uuid) -> bool {
        return land::PLand::getInstance().getLandRegistry()->isOperator(uuid);
    });

    exportAs("LandRegistry_addOperator", [](std::string const& uuid) -> bool {
        return land::PLand::getInstance().getLandRegistry()->addOperator(uuid);
    });

    exportAs("LandRegistry_removeOperator", [](std::string const& uuid) -> bool {
        return land::PLand::getInstance().getLandRegistry()->removeOperator(uuid);
    });

    exportAs("LandRegistry_hasPlayerSettings", [](std::string const& uuid) -> bool {
        return land::PLand::getInstance().getLandRegistry()->hasPlayerSettings(uuid);
    });

    exportAs("LandRegistry_getPlayerSettings", [](std::string const& uuid) -> std::string {
        try {
            // struct -> json -> std::string
            auto settings = land::PLand::getInstance().getLandRegistry()->getPlayerSettings(uuid);
            auto j        = land::JSON::structTojson(*settings);
            return j.dump();
        } catch (...) {
            return {};
        }
    });

    exportAs("LandRegistry_setPlayerSettings", [](std::string const& uuid, std::string const& jsonStr) -> bool {
        try {
            // jsonStr -> json -> struct
            auto json = nlohmann::json::parse(jsonStr);
            auto set  = land::PlayerSettings{};
            land::JSON::jsonToStructNoMerge(json, set);
            return land::PLand::getInstance().getLandRegistry()->setPlayerSettings(uuid, std::move(set));
        } catch (...) {
            return false;
        }
    });


    exportAs("LandRegistry_hasLand", [](int id) -> bool {
        return land::PLand::getInstance().getLandRegistry()->hasLand(id);
    });

    exportAs(
        "LandRegistry_addOrdinaryLand",
        [](InternalLandAABB iaabb, bool is3D, std::string const& owner) -> std::vector<std::string> {
            if (iaabb[0].second != iaabb[1].second) {
                throw std::runtime_error("LandRegistry_addOrdinaryLand: Invalid AABB");
            }
            auto dimId = iaabb[0].second;
            auto aabb  = toCpp<land::LandAABB>(iaabb);
            aabb.fix();

            auto land   = land::Land::make(aabb, dimId, is3D, owner);
            auto result = land::PLand::getInstance().getLandRegistry()->addOrdinaryLand(land);

            std::vector<std::string> ret;
            ret.reserve(2);

            if (result.has_value()) {
                ret.push_back("success");
                ret.push_back(std::to_string(land->getId()));
            } else {
                ret.push_back("error");
                ret.push_back(std::to_string(static_cast<int>(result.error())));
            }

            return ret;
        }
    );

    exportAs("LandRegistry_addSubLand", [](int parentLandId, InternalLandAABB iaabb) -> std::vector<std::string> {
        auto registry  = land::PLand::getInstance().getLandRegistry();
        auto parentPtr = registry->getLand(parentLandId);
        if (!parentPtr) {
            throw std::runtime_error("LandRegistry_addSubLand: Invalid parent land id");
        }
        auto aabb = toCpp<land::LandAABB>(iaabb);
        aabb.fix();

        auto subLand = land::Land::make(aabb, parentPtr->getDimensionId(), true, parentPtr->getOwner());
        auto result  = registry->addSubLand(parentPtr, subLand);

        std::vector<std::string> ret;
        ret.reserve(2);

        if (result.has_value()) {
            ret.push_back("success");
            ret.push_back(std::to_string(subLand->getId()));
        } else {
            ret.push_back("error");
            ret.push_back(std::to_string(static_cast<int>(result.error())));
        }

        return ret;
    });

    exportAs("LandRegistry_getLand", [](int id) -> int {
        auto land = land::PLand::getInstance().getLandRegistry()->getLand(id);
        if (!land) return -1;
        return land->getId();
    });

    exportAs("LandRegistry_removeLand", [](int id) -> bool {
        return land::PLand::getInstance().getLandRegistry()->removeLand(id);
    });
    exportAs("LandRegistry_removeOrdinaryLand", [](int id) -> int {
        auto ptr    = land::PLand::getInstance().getLandRegistry()->getLand(id);
        auto result = land::PLand::getInstance().getLandRegistry()->removeOrdinaryLand(ptr);
        return result.has_value() ? -1 : static_cast<int>(result.error());
    });
    exportAs("LandRegistry_removeSubLand", [](int id) -> int {
        auto ptr    = land::PLand::getInstance().getLandRegistry()->getLand(id);
        auto result = land::PLand::getInstance().getLandRegistry()->removeSubLand(ptr);
        return result.has_value() ? -1 : static_cast<int>(result.error());
    });
    exportAs("LandRegistry_removeLandAndSubLands", [](int id) -> int {
        auto ptr    = land::PLand::getInstance().getLandRegistry()->getLand(id);
        auto result = land::PLand::getInstance().getLandRegistry()->removeLandAndSubLands(ptr);
        return result.has_value() ? -1 : static_cast<int>(result.error());
    });
    exportAs("LandRegistry_removeLandAndPromoteSubLands", [](int id) -> int {
        auto ptr    = land::PLand::getInstance().getLandRegistry()->getLand(id);
        auto result = land::PLand::getInstance().getLandRegistry()->removeLandAndPromoteSubLands(ptr);
        return result.has_value() ? -1 : static_cast<int>(result.error());
    });
    exportAs("LandRegistry_removeLandAndTransferSubLands", [](int id) -> int {
        auto ptr    = land::PLand::getInstance().getLandRegistry()->getLand(id);
        auto result = land::PLand::getInstance().getLandRegistry()->removeLandAndTransferSubLands(ptr);
        return result.has_value() ? -1 : static_cast<int>(result.error());
    });


    using LandList = std::vector<land::LandID>;
    exportAs("LandRegistry_getLands", []() -> LandList {
        auto     lands = land::PLand::getInstance().getLandRegistry()->getLands();
        LandList result;
        result.reserve(lands.size());
        std::transform(lands.begin(), lands.end(), std::back_inserter(result), [](auto& land) {
            return land->getId();
        });
        return result;
    });

    exportAs("LandRegistry_getLands1", [](int dimid) -> LandList {
        auto     lands = land::PLand::getInstance().getLandRegistry()->getLands(dimid);
        LandList result;
        result.reserve(lands.size());
        std::transform(lands.begin(), lands.end(), std::back_inserter(result), [](auto& land) {
            return land->getId();
        });
        return result;
    });

    exportAs("LandRegistry_getLands2", [](std::string const& uuid, bool includeShared) -> LandList {
        auto     lands = land::PLand::getInstance().getLandRegistry()->getLands(uuid, includeShared);
        LandList result;
        result.reserve(lands.size());
        std::transform(lands.begin(), lands.end(), std::back_inserter(result), [](auto& land) {
            return land->getId();
        });
        return result;
    });

    exportAs("LandRegistry_getLands3", [](std::string const& uuid, int dimid) -> LandList {
        auto     lands = land::PLand::getInstance().getLandRegistry()->getLands(uuid, dimid);
        LandList result;
        result.reserve(lands.size());
        std::transform(lands.begin(), lands.end(), std::back_inserter(result), [](auto& land) {
            return land->getId();
        });
        return result;
    });

    exportAs("LandRegistry_getLands4", [](std::vector<int> lds) -> LandList {
        LandList int64List;
        int64List.reserve(lds.size());
        std::transform(lds.begin(), lds.end(), std::back_inserter(int64List), [](int i) {
            return static_cast<land::LandID>(i);
        });

        auto lands = land::PLand::getInstance().getLandRegistry()->getLands(int64List);

        LandList result;
        result.reserve(lands.size());
        std::transform(lands.begin(), lands.end(), std::back_inserter(result), [](auto& land) {
            return land->getId();
        });

        return result;
    });

    exportAs("LandRegistry_getPermType", [](std::string const& uuid, int landID, bool ignoreOperator) {
        return static_cast<int>(land::PLand::getInstance()
                                    .getLandRegistry()
                                    ->getPermType(uuid, static_cast<land::LandID>(landID), ignoreOperator));
    });

    exportAs("LandRegistry_getLandAt", [](IntPos const& pos) -> int {
        auto land = land::PLand::getInstance().getLandRegistry()->getLandAt(pos.first, pos.second);
        if (!land) return -1;
        return land->getId();
    });

    exportAs("LandRegistry_getLandAt1", [](IntPos const& pos, int radius) -> LandList {
        auto     lands = land::PLand::getInstance().getLandRegistry()->getLandAt(pos.first, radius, pos.second);
        LandList result;
        result.reserve(lands.size());
        std::transform(lands.begin(), lands.end(), std::back_inserter(result), [](auto& land) {
            return land->getId();
        });
        return result;
    });

    exportAs("LandRegistry_getLandAt2", [](IntPos const& a, IntPos const& b) -> LandList {
        auto     lands = land::PLand::getInstance().getLandRegistry()->getLandAt(a.first, b.first, a.second);
        LandList result;
        result.reserve(lands.size());
        std::transform(lands.begin(), lands.end(), std::back_inserter(result), [](auto& land) {
            return land->getId();
        });
        return result;
    });

    exportAs("LandRegistry_refreshLandRange", [](int id) -> void {
        auto inst = land::PLand::getInstance().getLandRegistry();
        auto land = inst->getLand(id);
        if (land) inst->refreshLandRange(land);
    });

    exportAs("PLand_getVersionMeta", []() -> std::string {
        static struct {
            int         major    = PLAND_VERSION_MAJOR;
            int         minor    = PLAND_VERSION_MINOR;
            int         patch    = PLAND_VERSION_PATCH;
            int         build    = PLAND_VERSION_BUILD;
            std::string commit   = PLAND_COMMIT_HASH;
            bool        snapshot = PLAND_VERSION_SNAPSHOT;
            bool        release  = PLAND_VERSION_RELEASE;
            std::string version  = PLAND_VERSION_STRING;
        } meta;
        static auto res = land::JSON::structTojson(meta).dump();
        return res;
    });
}


} // namespace ldapi

File: ts/imports/LandRegistry.ts
import {
    importAs,
    isIntPos,
    LandID,
    LandPermType,
    Result,
    UUIDs,
} from "../ImportDef.js";
import { LandAABB } from "./LandAABB.js";
import { Land } from "./Land.js";

/**
 * @warning 请不要增加、删除 key，否则会导致反射失败
 */
export type PlayerSettings = {
    /** 是否显示进入领地提示 */ showEnterLandTitle: boolean;
    /** 是否持续显示底部提示 */ showBottomContinuedTip: boolean;
    /** 玩家语言 */ localeCode: string | "system" | "server";
};

interface VersionMeta {
    major: number;
    minor: number;
    patch: number;
    build: number;
    commit: string;
    snapshot: boolean;
    release: boolean;
    version: string;
}

enum StorageLayerErrorCode {
    Unknown = 0, // 未知错误
    STLMapError = 1, // STL Map 操作失败
    DBError = 2, // 数据库操作失败
    DataConsistencyError = 3, // 数据一致性错误

    // addLand
    InvalidLand = 100, // 无效的领地数据
    AssignLandIdFailed = 101, // 分配领地ID失败
    LandRangeIllegal = 102, // 领地范围不合法

    // removeLand
    LandTypeWithRequireTypeNotMatch = 200, // 领地类型与要求类型不匹配
}

export class LandRegistry {
    static IMPORTS = {
        LandRegistry_isOperator: importAs("LandRegistry_isOperator"),
        LandRegistry_addOperator: importAs("LandRegistry_addOperator"),
        LandRegistry_removeOperator: importAs("LandRegistry_removeOperator"),
        LandRegistry_hasPlayerSettings: importAs(
            "LandRegistry_hasPlayerSettings"
        ),
        LandRegistry_getPlayerSettings: importAs(
            "LandRegistry_getPlayerSettings"
        ),
        LandRegistry_setPlayerSettings: importAs(
            "LandRegistry_setPlayerSettings"
        ),
        LandRegistry_getLand: importAs("LandRegistry_getLand"),
        LandRegistry_removeLand: importAs("LandRegistry_removeLand"),
        LandRegistry_hasLand: importAs("LandRegistry_hasLand"),
        LandRegistry_getLands: importAs("LandRegistry_getLands"),
        LandRegistry_getLands1: importAs("LandRegistry_getLands1"),
        LandRegistry_getLands2: importAs("LandRegistry_getLands2"),
        LandRegistry_getLands3: importAs("LandRegistry_getLands3"),
        LandRegistry_getLands4: importAs("LandRegistry_getLands4"),
        LandRegistry_getPermType: importAs("LandRegistry_getPermType"),
        LandRegistry_getLandAt: importAs("LandRegistry_getLandAt"),
        LandRegistry_getLandAt1: importAs("LandRegistry_getLandAt1"),
        LandRegistry_getLandAt2: importAs("LandRegistry_getLandAt2"),
        LandRegistry_refreshLandRange: importAs(
            "LandRegistry_refreshLandRange"
        ),
        PLand_getVersionMeta: importAs("PLand_getVersionMeta"),
        LandRegistry_removeOrdinaryLand: importAs(
            "LandRegistry_removeOrdinaryLand"
        ),
        LandRegistry_removeSubLand: importAs("LandRegistry_removeSubLand"),
        LandRegistry_removeLandAndSubLands: importAs(
            "LandRegistry_removeLandAndSubLands"
        ),
        LandRegistry_removeLandAndPromoteSubLands: importAs(
            "LandRegistry_removeLandAndPromoteSubLands"
        ),
        LandRegistry_removeLandAndTransferSubLands: importAs(
            "LandRegistry_removeLandAndTransferSubLands"
        ),
    };

    constructor() {
        throw new Error("PLand is a static class");
    }

    static isOperator(uuid: string): boolean {
        return LandRegistry.IMPORTS.LandRegistry_isOperator(uuid);
    }

    static addOperator(uuid: string): boolean {
        return LandRegistry.IMPORTS.LandRegistry_addOperator(uuid);
    }

    static removeOperator(uuid: string): boolean {
        return LandRegistry.IMPORTS.LandRegistry_removeOperator(uuid);
    }

    static hasPlayerSettings(uuid: string): boolean {
        return LandRegistry.IMPORTS.LandRegistry_hasPlayerSettings(uuid);
    }

    static getPlayerSettings(uuid: string): PlayerSettings | null {
        const jsonStr =
            LandRegistry.IMPORTS.LandRegistry_getPlayerSettings(uuid);
        try {
            return JSON.parse(jsonStr);
        } catch (e) {
            logger.error(`Failed to parse player settings: ${jsonStr}`);
            return null;
        }
    }

    static setPlayerSettings(uuid: string, settings: PlayerSettings): boolean {
        const jsonStr = JSON.stringify(settings);
        return LandRegistry.IMPORTS.LandRegistry_setPlayerSettings(
            uuid,
            jsonStr
        );
    }

    static hasLand(id: LandID): boolean {
        return LandRegistry.IMPORTS.LandRegistry_hasLand(id);
    }

    /**
     * 创建并添加一个普通领地(会触发事件)
     * @param aabb 领地范围
     * @param is3D 是否是3D领地
     * @param uuid 玩家UUID
     * @param orginalBuyPrice 原始购买价格(仅记录，用于后续删除领地的退款计算)
     * @native PLand::addLand()
     * @event PlayerBuyLandAfterEvent
     */
    static createAndAddLand(
        aabb: LandAABB,
        is3D: boolean,
        uuid: UUIDs,
        orginalBuyPrice: number = 0
    ): Land {
        if (aabb.min.dimid != aabb.max.dimid) {
            throw new Error("AABB min and max must be in the same dimension");
        }
        if (orginalBuyPrice < 0) {
            throw new Error("Original buy price must be non-negative");
        }
        // TODO: Call C++ API
        // TODO: 由于一些设计问题，此API先不实现，等 PLand 调整逻辑后实现
        throw new Error("Not implemented");
    }

    /**
     * @deprecated 此 API 在 PLand 底层中已标记为废弃函数，将在未来版本中移除，请使用 removeOrdinaryLand() 代替
     */
    static removeLand(landID: LandID): boolean {
        return LandRegistry.IMPORTS.LandRegistry_removeLand(landID);
    }

    /**
     * @brief 移除普通领地
     */
    static removeOrdinaryLand(
        land: Land | LandID
    ): Result<void, StorageLayerErrorCode> {
        const id =
            typeof land === "number"
                ? land
                : // @ts-ignore
                  (land as Land).unique_id;
        const res = LandRegistry.IMPORTS.LandRegistry_removeOrdinaryLand(id);
        return new Result<void, StorageLayerErrorCode>(
            res === -1 ? void 0 : null,
            res === -1 ? null : (res as StorageLayerErrorCode)
        );
    }

    /**
     * @brief 移除子领地
     */
    static removeSubLand(
        land: Land | LandID
    ): Result<void, StorageLayerErrorCode> {
        const id =
            typeof land === "number"
                ? land
                : // @ts-ignore
                  (land as Land).unique_id;
        const res = LandRegistry.IMPORTS.LandRegistry_removeSubLand(id);
        return new Result<void, StorageLayerErrorCode>(
            res === -1 ? void 0 : null,
            res === -1 ? null : (res as StorageLayerErrorCode)
        );
    }

    /**
     * @brief 移除领地和其子领地
     */
    static removeLandAndSubLands(
        land: Land | LandID
    ): Result<void, StorageLayerErrorCode> {
        const id =
            typeof land === "number"
                ? land
                : // @ts-ignore
                  (land as Land).unique_id;
        const res = LandRegistry.IMPORTS.LandRegistry_removeLandAndSubLands(id);
        return new Result<void, StorageLayerErrorCode>(
            res === -1 ? void 0 : null,
            res === -1 ? null : (res as StorageLayerErrorCode)
        );
    }

    /**
     * @brief 移除当前领地并提升子领地为普通领地
     */
    static removeLandAndPromoteSubLands(
        land: Land | LandID
    ): Result<void, StorageLayerErrorCode> {
        const id =
            typeof land === "number"
                ? land
                : // @ts-ignore
                  (land as Land).unique_id;
        const res =
            LandRegistry.IMPORTS.LandRegistry_removeLandAndPromoteSubLands(id);
        return new Result<void, StorageLayerErrorCode>(
            res === -1 ? void 0 : null,
            res === -1 ? null : (res as StorageLayerErrorCode)
        );
    }

    /**
     * @brief 移除当前领地并移交子领地给当前领地的父领地
     */
    static removeLandAndTransferSubLands(
        land: Land | LandID
    ): Result<void, StorageLayerErrorCode> {
        const id =
            typeof land === "number"
                ? land
                : // @ts-ignore
                  (land as Land).unique_id;
        const res =
            LandRegistry.IMPORTS.LandRegistry_removeLandAndTransferSubLands(id);
        return new Result<void, StorageLayerErrorCode>(
            res === -1 ? void 0 : null,
            res === -1 ? null : (res as StorageLayerErrorCode)
        );
    }

    static getLand(landID: LandID): Land | null {
        const id = LandRegistry.IMPORTS.LandRegistry_getLand(landID);
        if (id === -1) {
            return null;
        }
        return new Land(id);
    }

    /**
     * 获取所有领地
     */
    static getLands(): Land[];
    /**
     * 批量获取领地
     * @param ids id 列表
     */
    static getLands(ids: LandID[]): Land[];
    /**
     * 获取指定维度的所有领地
     * @param dimid 维度 id
     */
    static getLands(dimid: number): Land[];
    /**
     * 获取玩家所有的领地
     * @param uuid 玩家 uuid
     * @param includeShared 是否包含其它玩家共享的领地(默认false)
     */
    static getLands(uuid: UUIDs, includeShared: boolean): Land[];
    /**
     * 获取玩家在指定维度所有的领地
     * @param uuid 玩家 uuid
     * @param dimid 维度 id
     */
    static getLands(uuid: UUIDs, dimid: number): Land[];

    static getLands(...args: any[]): Land[] {
        switch (args.length) {
            case 0: {
                // getLands() const;
                return LandRegistry.IMPORTS.LandRegistry_getLands().map(
                    (id: LandID) => new Land(id)
                );
            }
            case 1: {
                if (Array.isArray(args[0])) {
                    // getLands(std::vector<LandID> const& ids) const;
                    return LandRegistry.IMPORTS.LandRegistry_getLands4(
                        args[0]
                    ).map((id: LandID) => new Land(id));
                } else if (typeof args[0] === "number") {
                    // getLands(LandDimid dimid) const;
                    return LandRegistry.IMPORTS.LandRegistry_getLands1(
                        args[0]
                    ).map((id: LandID) => new Land(id));
                }
            }
            case 2: {
                const arg0 = args[0];
                if (typeof arg0 === "string") {
                    if (typeof args[1] === "boolean" || args[1] === undefined) {
                        // getLands(UUIDs const& uuid, bool includeShared = false) const;
                        const includeShared = args[1] ?? false;
                        return LandRegistry.IMPORTS.LandRegistry_getLands2(
                            arg0,
                            includeShared
                        ).map((id: LandID) => new Land(id));
                    } else {
                        // getLands(UUIDs const& uuid, LandDimid dimid) const;
                        return LandRegistry.IMPORTS.LandRegistry_getLands3(
                            arg0,
                            args[1]
                        ).map((id: LandID) => new Land(id));
                    }
                }
            }
        }
        throw new Error("Invalid arguments");
    }

    /**
     * 查询指定坐标的领地
     * @param pos 坐标
     */
    static getLandAt(pos: IntPos): Land | null;
    /**
     * 查询指定坐标半径内的领地
     * @param pos 坐标
     * @param radius 半径
     */
    static getLandAt(pos: IntPos, radius: number): Land[];
    /**
     * 查询 AABB 区域内的领地
     * @param pos1 min 坐标
     * @param pos2 max 坐标
     */
    static getLandAt(pos1: IntPos, pos2: IntPos): Land[];

    static getLandAt(
        pos1: IntPos,
        radius_or_pos2?: number | IntPos
    ): Land | null | Land[] {
        if (!isIntPos(pos1)) {
            throw new Error("Invalid arguments");
        }

        if (radius_or_pos2 === undefined) {
            // getLandAt(BlockPos const& pos, LandDimid dimid) const;
            const id = LandRegistry.IMPORTS.LandRegistry_getLandAt(pos1);
            if (id === -1) {
                return null;
            }
            return new Land(id);
        } else if (typeof radius_or_pos2 === "number") {
            // getLandAt(BlockPos const& center, int radius, LandDimid dimid) const;
            return LandRegistry.IMPORTS.LandRegistry_getLandAt1(
                pos1,
                radius_or_pos2
            ).map((id: LandID) => new Land(id));
        } else if (isIntPos(radius_or_pos2)) {
            // getLandAt(BlockPos const& pos1, BlockPos const& pos2, LandDimid dimid) const;
            if (pos1.dimid != (radius_or_pos2 as IntPos).dimid) {
                throw new Error("Invalid arguments");
            }
            return LandRegistry.IMPORTS.LandRegistry_getLandAt2(
                pos1,
                radius_or_pos2
            ).map((id: LandID) => new Land(id));
        } else {
            throw new Error("Invalid arguments");
        }
    }

    static getPermType(
        uuid: UUIDs,
        landID = 0,
        ignoreOperator = false
    ): LandPermType {
        return LandRegistry.IMPORTS.LandRegistry_getPermType(
            uuid,
            landID,
            ignoreOperator
        );
    }

    static refreshLandRange(land: Land): void {
        // @ts-ignore
        LandRegistry.IMPORTS.LandRegistry_refreshLandRange(land.unique_id);
    }

    static getVersionMeta(): VersionMeta {
        return JSON.parse(LandRegistry.IMPORTS.PLand_getVersionMeta());
    }
}

Object.freeze(LandRegistry.IMPORTS);
